src/mcmc/gibbs.jl

```diff
@@ -1,251 +1,528 @@
+    for index in 1:length(samplers)
+        # Take the inner step.
+        sampler_local = samplers[index]
+        state_local = states[index]
+        varnames_local = _maybevec(varnames[index])
+        vi, new_state_local = gibbs_step_inner(
+            rng, model, varnames_local, sampler_local, state_local, vi; kwargs...
+        )
+        states = Accessors.setindex(states, new_state_local, index)
+    end
+    return Transition(model, vi), GibbsState(vi, states)
+end
 
-        # Compute initial state.
-        _, state = DynamicPPL.initialstep(rng, model, local_spl, vi; kwargs...)
+"""
+    setparams_varinfo!!(model, sampler::Sampler, state, params::AbstractVarInfo)
 
-        # Update `VarInfo` object.
-        vi = gibbs_varinfo(model, local_spl, state)
+A lot like AbstractMCMC.setparams!!, but instead of taking a vector of parameters, takes an
+`AbstractVarInfo` object. Also takes the `sampler` as an argument. By default, falls back to
+`AbstractMCMC.setparams!!(model, state, params[:])`.
 
-        return state
-    end
+`model` is typically a `DynamicPPL.Model`, but can also be e.g. an
+`AbstractMCMC.LogDensityModel`.
+"""
+function setparams_varinfo!!(model, ::Sampler, state, params::AbstractVarInfo)
+    return AbstractMCMC.setparams!!(model, state, params[:])
+end
 
-    # Compute initial transition and state.
-    transition = Transition(model, vi)
-    state = GibbsState(vi, samplers, states)
+function setparams_varinfo!!(
+    model::DynamicPPL.Model, sampler::Sampler{<:MH}, state::VarInfo, params::AbstractVarInfo
+)
+    # The state is already a VarInfo, so we can just return `params`, but first we need to
+    # update its logprob.
+    # TODO(mhauru) Is this the right context to use?
+    return last(DynamicPPL.evaluate!!(model, params, DynamicPPL.leafcontext(model.context)))
+end
 
-    return transition, state
+function setparams_varinfo!!(
+    model::DynamicPPL.Model,
+    sampler::Sampler{<:ESS},
+    state::VarInfo,
+    params::AbstractVarInfo,
+)
+    # The state is already a VarInfo, so we can just return `params`, but first we need to
+    # update its logprob.
+    # TODO(mhauru) Is this the right context to use?
+    return last(DynamicPPL.evaluate!!(model, params, DynamicPPL.LikelihoodContext()))
 end
 
-# Subsequent steps
-function AbstractMCMC.step(
-    rng::AbstractRNG, model::Model, spl::Sampler{<:Gibbs}, state::GibbsState; kwargs...
+function setparams_varinfo!!(
+    model::DynamicPPL.Model, sampler::Sampler, state::TuringState, params::AbstractVarInfo
 )
-    # Iterate through each of the samplers.
-    vi = state.vi
-    samplers = state.samplers
-    states = map(samplers, spl.alg.iterations, state.states) do _sampler, iteration, _state
-        # Recompute `vi.logp` if needed.
-        if _sampler.selector.rerun
-            vi = last(DynamicPPL.evaluate!!(model, rng, vi, _sampler))
-        end
+    logdensity = DynamicPPL.setmodel(state.logdensity, model, sampler.alg.adtype)
```

Comment:
Does `alg.adtype` always exist here?

Reply:
Now that the argument is type bound to `Sampler{<:ExternalSampler}`, yes.